### PR TITLE
Refactor umbrella for v2 ☂️☂️☂️

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
Add tests, refactor ambiguous code, improve performance 

## Problems
- Current tests are still based on the gotrue netlify's upstream logic - admins are identified based on the `is_super_admin` field in the `users` table. We should use the `role` to define an admin user.

## Things to fix
- [x] create admin user for testing `/admin` & `/audit` endpoints using `GOTRUE_JWT_ADMIN_ROLE` instead of setting `is_super_admin` field to `true`
- [x] `audit_log_entry` should identify admin user who makes the action (either service_role or supabase_admin) 
- [x] admin_test
- [x] signup_test
- [x] recovery_test
- [x] invite_test
- [x] hooks test
- [x] middleware test
- [x] external oauth provider tests (each provider creates a mock server to simulate the oauth2 flow)
- [x] instance test (multi-instance mode)

### CODE REVIEW

**Comment by ph1p** [comment]
I could not start postgres locally and had to do the following in the `/hack/migrate_postgres.sh` file:

```
-export GOTRUE_DB_DATABASE_URL="postgres://postgres:root@localhost:5432/gotrue_$DB_ENV?sslmode=disable"
+export GOTRUE_DB_DATABASE_URL="postgres://postgres:root@tcp(localhost:5432)/gotrue_$DB_ENV?sslmode=disable&parseTime=true&multiStatements=true"
```

Maybe some users also have the problem and you want to put it in your PR.

File: api/admin_test.go [hunk 1]
```diff
@@ -403,21 +408,44 @@ func (ts *AdminTestSuite) TestAdminUserDelete() {
...
-	}))
-
-	// Setup request
-	w := httptest.NewRecorder()
-	req := httptest.NewRequest(http.MethodPost, "/admin/users", &buffer)
-
-	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", ts.token))
-
-	ts.Config.External.Email.Enabled = false
+func (ts *AdminTestSuite) TestAdminUserCreateWithDisableLogin() {
```
Comment (thebengeu) [inline api/admin_test.go:411]: *DisabledLogin?

File: api/admin_test.go [hunk 2]
```diff
@@ -403,21 +408,44 @@ func (ts *AdminTestSuite) TestAdminUserDelete() {
...
+		require.NoError(ts.T(), json.NewEncoder(&buffer).Encode(c.userData))
+
+		// Setup request
+		w := httptest.NewRecorder()
+		req := httptest.NewRequest(http.MethodPost, "/admin/users", &buffer)
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", ts.token))
+
+		// Set config
+		*c.n = false
+		if i == 2 {
```
Comment (thebengeu) [inline api/admin_test.go]: Using index seems odd, having another field in `cases` would be more readable?

File: api/admin_test.go [hunk 3]
```diff
@@ -403,21 +408,44 @@ func (ts *AdminTestSuite) TestAdminUserDelete() {
...
+		}, http.StatusBadRequest},
+		{&ts.Config.DisableSignup, map[string]interface{}{
+			"email":    "test1@example.com",
+			"password": "test1",
+		}, http.StatusForbidden},
+	}
 
-	ts.API.handler.ServeHTTP(w, req)
-	require.Equal(ts.T(), http.StatusBadRequest, w.Code)
+	for i := 0; i < len(cases); i++ {
```
Comment (thebengeu) [inline api/admin_test.go:447]: Use for range loop?

File: api/api.go [hunk 4]
```diff
@@ -259,17 +258,18 @@ func (a *API) getConfig(ctx context.Context) *conf.Configuration {
...
-	smtpConfig := (*a.config).SMTP
-	if err := mergo.MergeWithOverwrite(&smtpConfig, config.SMTP); err != nil {
-		return nil
-	}
-	config.SMTP = smtpConfig
+	// smtpConfig := (*a.config).SMTP
+	// if err := mergo.MergeWithOverwrite(&smtpConfig, config.SMTP); err != nil {
+	// 	return nil
+	// }
+	// config.SMTP = smtpConfig
```
Comment (thebengeu) [inline api/api.go:271]: Could there be a comment on why this is commented out instead of removed entirely?

File: api/signup_test.go [hunk 5]
```diff
@@ -153,7 +153,7 @@ func (ts *SignupTestSuite) TestWebhookTriggered() {
 
 	w := httptest.NewRecorder()
 	ts.API.handler.ServeHTTP(w, req)
-	assert.Equal(http.StatusOK, w.Code)
+	// assert.Equal(http.StatusOK, w.Code)
```
Comment (thebengeu) [inline api/signup_test.go:156]: Is commenting out intended?

File: api/invite_test.go [hunk 6]
```diff
@@ -117,68 +116,64 @@ func (ts *InviteTestSuite) TestInvite_WithoutAccess() {
...
+			http.StatusOK,
+		},
+		{
+			"Verify invite with no password",
+			"test1@example.com",
+			map[string]interface{}{
+				"type":  "invite",
+				"token": "asdf",
+			},
+			http.StatusOK,
```
Comment (thebengeu) [inline api/invite_test.go:142]: How come this isn't http.StatusUnprocessableEntity?

**Review by thebengeu — APPROVED** [review]

File: api/signup_test.go (hunk 5 above)
Comment (kangmingtay) [reply api/signup_test.go:156]: whoops, nope this wasn't intended, thanks for catching this!

File: api/invite_test.go (hunk 6 above)
Comment (kangmingtay) [reply api/invite_test.go:142]: according to the `signupVerify` method, if the invited user doesn't have a password set, it still signs the user up with a random generated password (https://github.com/supabase/gotrue/blob/ceba0d5991b5230f25187baca9592cce576c5ca3/api/verify.go#L169). This probably maps to the invite button on the dashboard where admins can just invite a user by just typing in an email.

File: api/api.go (hunk 4 above)
Comment (kangmingtay) [reply api/api.go:271]: oh there's a comment on line 261 to explain what it does, but i'll make it more explicit to explain why it's commented out

Comment (kangmingtay) [reply api/api.go]: realised i don't have to comment it out in the end, just needed to check if it's in multi-instance mode before deciding whether to merge the respective configs

**Review by awalias — APPROVED** [review]
looks great!! will probably need a fair bit of end-to-end integration testing with so many changes (even with the new tests!)

**Comment by kiwicopple** [comment]
Amazing work here @kangmingtay 🔥

**Pull request merged**